### ISSUE TITLE
Breaking changes in Component structure(Dill)

### ISSUE THREAD

**Issue by kangwang1988** [dart-lang/sdk#39066]
This tracker is for issues related to:
* Common Front End (CFE) and kernel

I'm developing several packages based on kernel to kernel transformation. Including but not limited to AOP: (https://github.com/alibaba-flutter/aspectd)

However, I'm suffering from kind of breaking changes in dill structure in release/profile mode.
Now the frontend_server generates a dill file with commands below:
```shell
/Users/kylewong/Codes/Flutter/flutter/bin/cache/dart-sdk/bin/dart /Users/kylewong/Codes/Flutter/flutter/bin/cache/artifacts/engine/darwin-x64/frontend_server.dart.snapshot --sdk-root /Users/kylewong/Codes/Flutter/flutter/bin/cache/artifacts/engine/common/flutter_patched_sdk_product/ --strong --target=flutter -Ddart.developer.causal_async_stacks=true -Ddart.vm.profile=false -Ddart.vm.product=true --bytecode-options=source-positions --aot --tfa --packages .packages --output-dill build/aot/app.dill --depfile build/aot/kernel_compile.d package:example/main.dart
```
It generates a dill file of bytecode without ast structure. Henceforth, the old logic I written based on the tree structure from Component->Library->Class->Method->Statements can only works for debug mode, and would fail in release/profile mode.

Now I can rewrite the dill generate logic by adding flags in flutter_tools,
```shell
--sdk-root /Users/kylewong/Codes/Flutter/flutter/bin/cache/artifacts/engine/common/flutter_patched_sdk_product/ --strong --target=flutter -Ddart.developer.causal_async_stacks=true -Ddart.vm.profile=false -Ddart.vm.product=true --bytecode-options=source-positions --aot --tfa --packages .packages --output-dill build/aot/app.dill --depfile build/aot/kernel_compile.d package:example/main.dart --no-drop-ast --no-gen-bytecode
```
--no-drop-ast --no-gen-bytecode will do the trick.

I have several questions:
a. Is it impossible for me to parse the Component with tree structure of Component->Library->Class->Method->Statements when -drop-ast=true?
b. When I add --no-drop-ast only in order to contain the ast information in the dill file, the frontend_server will fail with errors below:
<img width="1674" alt="Screen Shot 2019-10-23 at 9 47 28 PM" src="https://user-images.githubusercontent.com/817851/67399579-09a79300-f5df-11e9-9dc4-33baeee60b25.png">
Is it by design? I want to modify the dill generate process as little as possible.
c. In the perspective of generated app-aot-blobs/assembly/elf, will "--no-drop-ast --no-gen-bytecode" make any difference?

My flutter doctor -v
```shell
[✓] Flutter (Channel master, v1.10.15-pre.218, on Mac OS X 10.15 19A602, locale
    en-CN)
    • Flutter version 1.10.15-pre.218 at /Users/kylewong/Codes/Flutter/flutter
    • Framework revision a817b4e6a4 (7 hours ago), 2019-10-23 03:14:06 -0400
    • Engine revision f0a2850b35
    • Dart version 2.6.0 (build 2.6.0-dev.8.1 17ed3cb5b3)
```

**Comment by kangwang1988** [comment]
@a-siva @mraleph

**Comment by mraleph** [comment]
I think we need to change Kernel Front-end in a way that makes it extensible - so that you can run your Kernel-to-Kernel transformation before bytecode generation happens. Then bytecode generator could transform Kernel AST after you mutated it. 

/cc @mkustermann @alexmarkov Alex, Martin - any opinions on this? Any suggestions how we could add an extension point into the process?

**Comment by mkustermann** [comment]
> /cc Alex, Martin - any opinions on this?
> Any suggestions how we could add an extension point into the process?

The bytecode-based Kernel is effectively a VM-specific file format.

We probably should allow our tools to emit the AST-kernel format to disc and have a separate tool which can consume AST-kernel and produce the VM-specific bytecode format.

It doesn't have to be the default way the tools work, but we should certainly make it possible without too much hassle.

That allows 3rd party code to perform arbitrary program analysis, transformations and does not tightly couple it to our CFE / frontend server / ... It also allows users e.g. to dump the AST based kernel after global transformations and examine it in textual form or programmatically.


A further step would be to allow injecting custom logic into the the CFE at various stages. Dynamically loading custom code into the CFE-running isolate or communicating across isolate/process boundaries might not be the desired solution. It creates tight coupling and potentially versioning issues.
Maybe a better way would be to allow users to make their own frontend-server based on components we provide?

**Comment by kangwang1988** [comment]
Also related to: https://github.com/flutter/flutter/issues/36738

**Comment by alexmarkov** [comment]
@kangwang1988 If you're modifying frontend_server, you can add your transformation before bytecode generation: https://github.com/dart-lang/sdk/blob/669375b269dfd3d628f29c129d2acc13f88463df/pkg/vm/lib/kernel_front_end.dart#L320-L322

This would allow you to transform kernel AST before it is replaced with bytecode.

Even better, I would suggest to add it before global transformations which include tree shaking:
https://github.com/dart-lang/sdk/blob/669375b269dfd3d628f29c129d2acc13f88463df/pkg/vm/lib/kernel_front_end.dart#L308-L309

This will make sure that results of your transformation are properly tree-shaken.

@mkustermann We can make a tool which would translate AST .dill file into bytecode. However, this would not cover the case of incremental compilation (in Flutter debug mode), where we create dill files on the fly. So I think we should have some extension points in frontend_server which could be used to inject custom transformations. Widget tracking transformation was using the following hook:

https://github.com/dart-lang/sdk/blob/669375b269dfd3d628f29c129d2acc13f88463df/pkg/frontend_server/lib/frontend_server.dart#L255-L257

https://github.com/dart-lang/sdk/blob/669375b269dfd3d628f29c129d2acc13f88463df/pkg/frontend_server/lib/frontend_server.dart#L442

Since then, widget tracking was moved to earlier stages of front-end (in order to happen before constant evaluation), so this transformation callback is probably outdated.

We can provide such callbacks to transform AST Component right after front-end (before global transformations), and before bytecode generation. In such case, transformations can be injected via the customized main for frontend_server which would pass transformations to the common frontend_server entry point. Flutter tools can be modified to switch to customized frontend_server if certain option (or pubsepc.yaml directive) is specified.

As for the tight coupling, I don't think it can be avoided. AST changes over time (mainly due to new language features), and transformation passes are inherently tightly coupled with ast.dart API. Custom transformation should use the same version of ast.dart and, in case of a separate tool it should also use the same version of AST reader and writer.

**Comment by mkustermann** [comment]
> Widget tracking transformation was using the following hook ...

That particular code is under our control. This discussion is about arbitrary 3rd party code.


> However, this would not cover the case of incremental compilation (in Flutter debug mode), where 
> we create dill files on the fly. So I think we should have some extension points in frontend_server 
> which could be used to inject custom transformations. 
> ...

Yes, the question is how to "inject" custom transformations.

As I suggested above I would vote for enabling users be able to write their own frontend server (or having a tool to generate one) - using the building blocks we already have - and letting `flutter_tools` talk to such 3rd party frontend servers (e.g. instead of having only our front-end server and letting it dynamically load code (there is some prior experience with barback in this regard)).

It would require thinking at least about:

  * **Distribution**: The custom front-end server / custom transformations need to be able to import our building blocks (e.g. `package:kernel`, `package:vm`, `package:frontend_server`, ...). The question is how to distribute those building blocks and make them accessible:

    - The normal way would be to publish them on pub. Due to the instability of AST and bytecode it would require us to publish them very frequently with very narrow sdk constraints. So it might not be the ideal solution.

    - The alternative would be to distribute them as part of our `dart-sdk`. Right now it is not possible to make a package depend on sources inside the sdk in the `packages.yaml` file, but that could be added.

  * **Versioning**: The custom transformations will live in packages. Those packages will necessarily be tied closely to the building blocks. So the packages containing custom transformations will need to be able to specify constraints on the building blocks (which might be inside the sdk, i.e. sdk constraint).

  * **Ordering**: Having multiple custom transformations will require an ordering between them. A custom front-end server can specify that in code. It would be also feasible to generate a custom front-end server, in which case e.g. a yaml file could specify the list of custom transformers with their import path.

  * **Tooling**: The `flutter_tools` would need to use a custom frontend server. One could even argue such a mechanism should work outside flutter as well, which would imply that the VM should be able to use a custom kernel-service when running dart from source. The path to such a custom front end server / kernel-service could be encoded in the yaml files

**Comment by kangwang1988** [comment]
cc @jonahwilliams

**Comment by alexmarkov** [comment]
> b. When I add --no-drop-ast only in order to contain the ast information in the dill file, the frontend_server will fail with errors below:
> Screen Shot 2019-10-23 at 9 47 28 PM
> Is it by design? I want to modify the dill generate process as little as possible.

Fixed in 8a09d7ab5ad4beafeecec3209d7a9d952e2d69be.

**Comment by kangwang1988** [comment]
I'm closing this issue as I've refactor aspectd so that it can work like the widgettrackcreation, where i don't need to read and write dill directly.
See: https://github.com/alibaba-flutter/sdk/commit/9275718097165f192b49ca2eb03547e04f2f4085 for more.
